notebooks/implicit_differentiation.ipynb

```diff
@@ -0,0 +1,561 @@
+    "\n",
+    "There is only one thing we have to change besides the normal stuff we have to change when using neighbor lists. We simply add the ```has_aux=True``` keyword to ```implicit_diff.custom_root```."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "massive-formation",
+   "metadata": {},
+   "source": [
+    "There's one more funny thing to note about the neighbor list version. Instead of getting an ```CustomVJPException``` when we forget to wrap our ```force_fn``` with ```lax.stop_gradient``` we get no explicit error message but we now get a memory leak instead which drastically slows down the computation. So pay attention that your are passing the correct version of your ```force_fn``` to your ```solver```!"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "consecutive-adapter",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# We add the parameter forced_rebuilding to make sure \n",
+    "# that it is possible to construct new neighbor lists during the optimization.\n",
+    "\n",
+    "def implicit_diff_nl(params,R_init,box_size,forced_rebuilding):\n",
+    "    neighbor_fn, energy_fn = energy.soft_sphere_neighbor_list(displacement, box_size, **params)\n",
+    "\n",
+    "    force_fn = jit(quantity.force(energy_fn))\n",
+    "    \n",
+    "    # wrap force_fn with a lax.stop_gradient to prevent a memory leak\n",
+    "    no_grad_force_fn = jit(lambda x,neighbor: lax.stop_gradient(force_fn(x,neighbor)))\n",
+    "    \n",
+    "    # make the dependence on the variables we want to differentiate explicit\n",
+    "    explicit_force_fn = jit(lambda R, neighbor, p: force_fn(R,neighbor,**p))\n",
+    "    \n",
+    "    def solver(params, x):\n",
+    "        # params are unused\n",
+    "        del params\n",
+    "        # need to use no_grad_force_fn!\n",
+    "        return run_minimization_while_nl(neighbor_fn,no_grad_force_fn, x, forced_rebuilding, shift, dt_start=0.001, dt_max=0.005)\n",
+    "    \n",
+    "    # Need to use hax_aux=True\n",
+    "    decorated_solver = jaxopt.implicit_diff.custom_root(explicit_force_fn,has_aux=True)(solver)\n",
+    "    \n",
+    "    R_final, nbrs, num_steps = decorated_solver(None,R_init)\n",
+    "    \n",
+    "    # Here we can just use our original energy_fn/force_fn\n",
+    "    return energy_fn(R_final,neighbor=nbrs), jnp.amax(jnp.abs(force_fn(R_final,neighbor=nbrs)))\n",
+    "\n",
+    "(imp_nl_e, imp_nl_f), imp_nl_g = jax.value_and_grad(implicit_diff_nl,has_aux=True)(param_dict,R_init,box_size,True)\n",
+    "print('Energy        : ',imp_nl_e)\n",
+    "print('Max_grad_force: ',imp_nl_f)\n",
+    "print('Gradient of the energy:')\n",
+    "print(imp_nl_g['sigma'][0])\n",
+    "print(imp_nl_g['alpha'][0])"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "behind-clark",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# We again get the same results.\n",
```

Comment:
Would it be worth adding one more section where you use a much larger system to show how well the neighbor list version scales?

Reply:
The only reason I added the neighbor list version in the first place was to demonstrate that it works in the first place as I had my own version of implicit diff working for quite some time except for the case where I tried to combine it with neighbor lists.

If you want to I can add a larger example. But I'd assume that people already know what neighbor lists are good for when they try to use implicit differentiation.